lib/cypress/record_filter.rb

```diff
@@ -0,0 +1,94 @@
+module Cypress
+  class RecordFilter
+    def self.filter(records, filters, options)
+      records.where create_query(filters, options)
+    end
+
+    def self.create_query(input_filters, options)
+      query = {}
+
+      if input_filters['races']
+        query['race.code'] = { '$in' => input_filters['races'] }
+      end
+
+      if input_filters['ethnicities']
+        query['ethnicity.code'] = { '$in' => input_filters['ethnicities'] }
+      end
+
+      if input_filters['genders']
+        query['gender'] = { '$in' => input_filters['genders'] }
+      end
+
+      if input_filters['payers']
+        query['insurance_providers'] = { '$elemMatch' => { 'payer.name' => { '$in' => input_filters['payers'] } } }
+      end
+
+      if input_filters['age']
+        create_age_query(query, input_filters['age'], options)
+      end
+
+      if input_filters['problems']
+        create_problem_query(query, input_filters['problems'])
+      end
+
+      # STILL TODO:
+      # provider tin
+      # provider npi
+      # provider type
+      # practice site addr
+
+      query
+    end
+
+    def self.create_age_query(query, age_filter, options)
+      # filter only by a single age range, can be age < max, age > min, or min < age < max
+      effective_date = options[:effective_date]
+
+      if age_filter['max']
+        age_max = age_filter['max']
+        req_birthdate = Time.at(effective_date).utc - age_max.years
+
+        query[:birthdate.gte] = req_birthdate
+      end
```

Comment:
If following the measures conventions of what age at measure period means this needs to account for some corner case issues.   For instance , if the filter is 18 year at Measure Period then due to the way date time logic is performed in the measures someone who is born on 01/01/xxxx 18 years from the MP still meets the measure logic even though the absolute value for the birth time compared to the start of the measure period would state otherwise.